File: insert-interval.ts
{
    function merge(intervals: number[][]): number[][] {
        intervals.sort((a,b) => a[0] - b[0]);
        
        const result: number[][] = [];
    
        let last = intervals[0];
    
        for (let i = 1; i < intervals.length; i++) {
    
            if (last[1] >= intervals[i][0]) {
                last[1] = Math.max(last[1], intervals[i][1]);         
            } else {
                result.push(last)
                last = intervals[i];
            }
    
    
        }
    
    
        return result.concat([last])
    };

    function insert(intervals: number[][], newInterval: number[]): number[][] {
        intervals.push(newInterval);

        return merge(intervals)
    };
}

File: insert-interval.rs
pub fn insert(mut intervals: Vec<Vec<i32>>, mut new_int: Vec<i32>) -> Vec<Vec<i32>> {
    let i = intervals.partition_point(|v| v[1] < new_int[0]);
    let j = intervals.partition_point(|v| v[0] <= new_int[1]);

    let slice: Vec<_> = intervals.drain(i..j).flatten().collect();

    if let [first, .., last] = &slice[..] {
        new_int[0] = new_int[0].min(*first);
        new_int[1] = new_int[1].max(*last);
    }

    intervals.insert(i, new_int);

    intervals
}
